lib/python/qmk/keymap.py

```diff
@@ -189,6 +189,43 @@ def generate_json(keymap, keyboard, layout, layers):
     return new_keymap
 
 
+def add_layout_data(new_keymap, keymap_json):
+    layout_macro = keymap_json['layout']
+
+    if keymap_json.get('layout_data'):
+        for (data_name, data_val) in keymap_json.get('layout_data').items():
+            data_val = map(_remove_unsafe_chars, data_val)
+            data_name_txt = '__REPLACE_%s__' % _remove_unsafe_chars(data_name)
+            new_keymap = new_keymap.replace(data_name_txt, '%s(%s)' % (layout_macro, ', '.join(data_val)))
+
+    if keymap_json.get('layer_data'):
+        for (data_name, data_val) in keymap_json.get('layer_data').items():
+            layer_data_txt = []
+            for layer_num, layer_data in enumerate(data_val):
+                if len(layer_data_txt) > 0:
+                    layer_data_txt[-1] = layer_data_txt[-1] + ','
+                if type(layer_data) is list:
+                    layer_data = map(_remove_unsafe_chars, layer_data)
+                    layer_data_txt.append('\t[%s] = %s(%s)' % (layer_num, layout_macro, ', '.join(layer_data)))
+                else:
+                    layer_data_txt.append('\t' + _remove_unsafe_chars(layer_data))
+            data_name_txt = '__REPLACE_%s__' % _remove_unsafe_chars(data_name)
+            new_keymap = new_keymap.replace(data_name_txt, '\n'.join(layer_data_txt))
+
+    if keymap_json.get('layer_custom_keys'):
+        layer_keys_txt = []
+        c_macro = keymap_json['layout'] + '_CUSTOM_KEYS'
```

Comment:
All LGTM, and I'm thinking if QMK maintainers would accept these changes.

Other `REPLACE_$blah` pattern looks LGTM as those are easily customizable by other keyboards as well. I'd assume layer count is also usable by other keyboards. So these 2 changes should be easier to be accepted by QMK maintainers.

But I'm not sure if this `_CUSTOM_KEYS_` key thing is extensible enough to be accepted by them 🤔 I'm thinking can we make a general enough routine that can be used by both `key_map` and `custom_keymap`. Can we make `CUSTOM` replaceable like those in `REPLACE_$blah`? So the code here matches `$blah` to `$BLAH_KEYS_GOES_HERE`

Reply:
Good points. But let's get this working first before adjusting the code so they are acceptable to maintainers. If I was the maintainer, there are a few things I'd want to change before accepting the change. _CUSTOM_KEYS_ can be controversial - but we need to get it working now so let's go forward. We can discuss with maintainers later how best to make that work.